README.md
# Osmolytes and protein stability

![pytest](https://github.com/Electrostatics/osmolytes/workflows/pytest/badge.svg)

This code attempts to predict the influence of osmolytes on protein stability, using the methods in:  Auton M, Bolen DW. Predicting the energetics of osmolyte-induced protein folding/unfolding. _Proc Natl Acad Sci_ 102:15065 (2005) https://doi.org/10.1073/pnas.0507053102

Other models may be added in the future.

## Installation

Regardless of how you install the software, we recommend doing so in a virtual environment. 
You can create a virtual environment using, e.g., one of the following:

```
conda create -n osmolytes pip
```
(note that you will need the `pip` package for later steps) or
```
virtualenv -p python3 ~/venvs/osmolytes
```

You will need to activate the virtual environment prior to installing or using software, e.g., with one of the following:

```
conda activate osmolytes
```
or
```
. ~/venvs/osmolytes/bin/activate
```

### Install from PyPI repository

Most users will want to install the software from the [PyPI repository](https://pypi.org/project/osmolytes/0.0.1/) via the following steps:

1. Activate your virtual environment as described above
2. Install the package using
```
pip install osmolytes
```

### Install from source code

Adventurous users and developers may want to install the software from source code via the following steps:

1. Activate your virtual environment as described above
2. Download the source code
```
git clone https://github.com/Electrostatics/osmolytes.git
```
3. Change to the source code directory (e.g., `cd osmolytes`)
4. Install the package from source code using
```
pip install -e .
```
If you are using conda, you may first need to install pip with
```
conda install pip
```

Installing from source also allows testing of the software with 
```
python -m pytest .
```
These tests check the integrity of the software and provide information about the accuracy of the various calculations (particularly, surface area).

## Usage

The main command-line tool installed by this package is the program `mvalue`.

The input to the `mvalue` program is a [PQR file](https://pdb2pqr.readthedocs.io/en/latest/formats/pqr.html) which can be generated by the [PDB2PQR software](https://pdb2pqr.readthedocs.io/en/latest/index.html). 
Most users will want to use the [PDB2PQR web server](http://server.poissonboltzmann.org/).
The `mvalue` program has been tested with PQR files generated by PDB2PQR using the following settings:

* Protonation states assigned by PROPKA
* PARSE force field
* Internal naming scheme

In addition to generating a PQR, the [PROPKA](https://github.com/jensengroup/propka) software generates additional output in the form of `*.propka` files that contain useful predictions about the pH stability of the protein.

The `mvalue` program has several options which can be listed using 
```
mvalue --help
```
The basic usage of `mvalue` is 
```
mvalue 2BU4.pqr
```
where `2BU4.pqr` is the name of an example PQR file.

The `mvalue` program produces output that describes 

* The residue composition of the protein
* The contribution of each residue type (and its surface area) to the *m*-value of the protein for each osmolyte species (at 1M concentration).
* The *m*-value associated with each osmolyte at 1M concentration.

The concept of a *m*-value is described in [Wikipedia](https://en.wikipedia.org/wiki/Equilibrium_unfolding) and many review articles on protein folding and stability.

# Support

Re-implementation of these algorithms was supported by National Institutes of Health grant GM069702.



CHANGELOG.md
# osmolytes

## 1.2.0 (17-Dec-2020)

## Additions

* Output surface area per residue type

## 1.1.1 (17-Dec-2020)

### Fixes

* Remove excess output on stdout/stderr

## 1.1.0 (16-Dec-2020)

### Additions

* CSV and Excel output

### Changes

* Refactored tests to also test main driver (where relevant)

## 1.0.0 (15-Dec-2020)

### Changes

* Restricted Python version to 3.8+.

### Fixes

* Removed [versioneer](https://github.com/python-versioneer/python-versioneer) support for versioning due to lack of compatibility with PEP517.
Replaced with `importlib` suggestion from [PyPA](https://packaging.python.org/guides/single-sourcing-package-version/).

## 0.1.0

### Additions

* Add [versioneer](https://github.com/python-versioneer/python-versioneer) support for versioning

### Fixes

Fixed problem with insertion codes and added tests for problems in the future.

## 0.0.1 (26-Nov-2020)

Initial release.

tests/test_pqr.py
"""Test PQR parsing."""
import logging
import pytest
from pathlib import Path
import yaml
import numpy as np
import pandas as pd
from osmolytes.pqr import parse_pqr_file, count_residues


_LOGGER = logging.getLogger(__name__)
PQR_DATA = [
    ["ATOM", 1, "N", "MET", "1", -22.600, 7.684, -9.315, -0.3200, 2.0000],
    ["ATOM", 2, "CA", "MET", "1", -21.455, 8.620, -9.099, 0.3300, 2.0000],
    ["ATOM", 3, "C", "MET", "1", -21.898, 10.062, -8.899, 0.5500, 1.7000],
    ["ATOM", 4, "O", "MET", "1", -22.362, 10.721, -9.827, -0.5500, 1.4000],
    ["ATOM", 5, "CB", "MET", "1", -20.471, 8.544, -10.268, 0.0000, 2.0000],
    ["ATOM", 6, "CG", "MET", "1", -19.558, 7.330, -10.209, 0.2650, 2.0000],
    ["ATOM", 7, "SD", "MET", "1", -18.624, 7.277, -8.646, -0.5300, 1.8500],
    ["ATOM", 8, "CE", "MET", "1", -16.958, 7.708, -9.204, 0.2650, 2.0000],
    ["ATOM", 9, "HE1", "MET", "1", -16.809, 7.326, -10.114, 0.0000, 0.0000],
    ["ATOM", 10, "HE2", "MET", "1", -16.871, 8.702, -9.236, 0.0000, 0.0000],
    ["ATOM", 11, "HE3", "MET", "1", -16.291, 7.330, -8.565, 0.0000, 0.0000],
    ["ATOM", 12, "H2", "MET", "1", -22.539, 6.944, -8.650, 0.3300, 0.0000],
    ["ATOM", 13, "H3", "MET", "1", -23.453, 8.186, -9.193, 0.3300, 0.0000],
    ["ATOM", 14, "HG2", "MET", "1", -20.120, 6.488, -10.314, 0.0000, 0.0000],
    ["ATOM", 15, "HG3", "MET", "1", -18.920, 7.356, -11.001, 0.0000, 0.0000],
    ["ATOM", 16, "H", "MET", "1", -22.546, 7.319, -10.241, 0.3300, 0.0000],
    ["ATOM", 17, "HA", "MET", "1", -20.974, 8.313, -8.264, 0.0000, 0.0000],
    ["ATOM", 18, "HB3", "MET", "1", -19.900, 9.372, -10.269, 0.0000, 0.0000],
    ["ATOM", 19, "HB2", "MET", "1", -20.989, 8.512, -11.130, 0.0000, 0.0000],
    ["ATOM", 20, "N", "GLN", "2", -21.741, 10.545, -7.674, -0.4000, 1.5000],
    ["ATOM", 21, "CA", "GLN", "2", -22.113, 11.910, -7.330, -0.0000, 2.0000],
    ["ATOM", 22, "C", "GLN", "2", -20.881, 12.616, -6.812, 0.5500, 1.7000],
    ["ATOM", 23, "O", "GLN", "2", -19.835, 11.997, -6.608, -0.5500, 1.4000],
    ["ATOM", 24, "CB", "GLN", "2", -23.203, 11.925, -6.251, 0.0000, 2.0000],
    ["ATOM", 25, "CG", "GLN", "2", -24.481, 11.246, -6.670, 0.0000, 2.0000],
    ["ATOM", 26, "CD", "GLN", "2", -25.494, 11.116, -5.541, 0.5500, 1.7000],
    ["ATOM", 27, "HG2", "GLN", "2", -24.905, 11.773, -7.411, 0.0000, 0.0000],
    ["ATOM", 28, "HA", "GLN", "2", -22.407, 12.370, -8.165, 0.0000, 0.0000],
    ["ATOM", 29, "HG3", "GLN", "2", -24.268, 10.325, -7.007, 0.0000, 0.0000],
    ["ATOM", 30, "H", "GLN", "2", -21.335, 9.873, -6.990, 0.4000, 1.0000],
    ["ATOM", 31, "HB3", "GLN", "2", -23.409, 12.879, -6.023, 0.0000, 0.0000],
    ["ATOM", 32, "HB2", "GLN", "2", -22.848, 11.462, -5.437, 0.0000, 0.0000],
    ["ATOM", 33, "HE22", "GLN", "2", -25.445, 9.103, -5.740, 0.3900, 1.0000],
    ["ATOM", 34, "OE1", "GLN", "2", -25.958, 12.106, -4.969, -0.5500, 1.4000],
    ["ATOM", 35, "NE2", "GLN", "2", -25.846, 9.881, -5.221, -0.7800, 1.5000],
    ["ATOM", 36, "HE21", "GLN", "2", -26.513, 9.707, -4.463, 0.3900, 1.0000],
    ["ATOM", 37, "N", "ARG", "3", -21.000, 13.920, -6.624, -0.4000, 1.5000],
    ["ATOM", 38, "CA", "ARG", "3", -19.885, 14.678, -6.101, -0.0000, 2.0000],
    ["ATOM", 39, "C", "ARG", "3", -20.291, 15.253, -4.763, 0.5500, 1.7000],
    ["ATOM", 40, "O", "ARG", "3", -21.384, 15.807, -4.617, -0.5500, 1.4000],
]
PROTEIN_PATH = Path("tests/data/proteins")


@pytest.mark.parametrize(
    "pqr_path,chain_id",
    [("NikR_no-chains.pqr", None), ("NikR_chains.pqr", "A")],
)
def test_parsing(pqr_path, chain_id):
    """Test PQR parsing with and without chains"""
    pqr_path = PROTEIN_PATH / pqr_path
    with open(pqr_path) as pqr_file:
        atoms = parse_pqr_file(pqr_file)
        for iatom in range(40):
            test_atom = atoms[iatom]
            ref_atom = PQR_DATA[iatom]
            assert test_atom.entry_type == ref_atom[0]
            assert test_atom.pqr_atom_num == ref_atom[1]
            assert test_atom.atom_name == ref_atom[2]
            assert test_atom.res_name == ref_atom[3]
            assert test_atom.res_num == ref_atom[4]
            assert test_atom.chain_id == chain_id
            ref_pos = np.array([ref_atom[5], ref_atom[6], ref_atom[7]])
            np.testing.assert_almost_equal(
                test_atom.position, ref_pos, decimal=3
            )
            np.testing.assert_almost_equal(
                test_atom.charge, ref_atom[8], decimal=3
            )
            np.testing.assert_almost_equal(
                test_atom.radius, ref_atom[9], decimal=3
            )


@pytest.mark.parametrize("protein", ["1A6F", "1STN", "2BU4"])
def test_residue_count(protein):
    """Test aggregate SASAs per residue type as reported in Auton and Bolen
    (doi:10.1073/pnas.0507053102, Supporting Table 2)."""
    pqr_path = PROTEIN_PATH / f"{protein}.pqr"
    with open(pqr_path, "rt") as pqr_file:
        atoms = parse_pqr_file(pqr_file)
    test_counts = count_residues(atoms).sort_index()
    ref_path = PROTEIN_PATH / "Auton-Bolen.yaml"
    with open(ref_path, "rt") as ref_file:
        ref_dict = yaml.load(ref_file, Loader=yaml.FullLoader)
    ref_dict = ref_dict[protein]
    ref_df = pd.DataFrame(ref_dict).T
    ref_df = ref_df[ref_df["number"] > 0]
    ref_counts = np.array(ref_df["number"].sort_index())
    results_df = pd.DataFrame({"Test": test_counts, "Ref": ref_counts})
    _LOGGER.info(f"Results for {protein}:\n{results_df}")
    np.testing.assert_equal(test_counts, ref_counts)


@pytest.mark.parametrize("protein", ["2PTN", "1OVA"])
def test_insertion_codes(protein):
    """Test whether proteins with insertion codes break the software."""
    pqr_path = PROTEIN_PATH / f"{protein}.pqr"
    with open(pqr_path, "rt") as pqr_file:
        parse_pqr_file(pqr_file)


tests/test_sasa.py
"""Test solvent-accessible surface area methods."""
import logging
import json
from pathlib import Path
import pytest
import yaml
from scipy.stats import linregress
import numpy as np
import pandas as pd
from osmolytes.sasa import SolventAccessibleSurface, ReferenceModels
from osmolytes.pqr import parse_pqr_file, Atom, aggregate, count_residues


_LOGGER = logging.getLogger(__name__)


with open("tests/data/alkanes/alkanes.json", "rt") as json_file:
    ATOM_AREAS = json.load(json_file)
PROTEIN_PATH = Path("tests/data/proteins")


@pytest.mark.parametrize("radius", [0.25, 0.5, 1.0, 2.0, 4.0])
def test_one_sphere_sasa(radius, tmp_path):
    """Test solvent-accessible surface areas for one sphere."""
    atom = Atom()
    atom.position = np.random.randn(3)
    frac = np.random.rand(1)[0]
    atom.radius = frac * radius
    probe_radius = (1.0 - frac) * radius
    xyz_path = Path(tmp_path) / "sphere.xyz"
    sas = SolventAccessibleSurface(
        [atom], probe_radius, 1000, xyz_path=xyz_path
    )
    atom_sasa = sas.atom_surface_area(0)
    ref_sasa = 4.0 * np.pi * radius * radius
    _LOGGER.info(
        f"Radius: {radius}, Test area: {atom_sasa}, Ref area: {ref_sasa}"
    )
    np.testing.assert_almost_equal(atom_sasa, ref_sasa)


def two_sphere_area(radius1, radius2, distance):
    """Area of two overlapping spheres.

    :param float radius1:  radius of sphere1
    :param float radius2:  radius of sphere2
    :param float distance:  distance between centers of spheres
    :returns:  exposed areas of spheres
    :rtype:  (float, float)
    """
    distsq = distance * distance
    rad1sq = radius1 * radius1
    rad2sq = radius2 * radius2
    full_area1 = 4 * np.pi * rad1sq
    full_area2 = 4 * np.pi * rad2sq
    if distance > (radius1 + radius2):
        return (full_area1, full_area2)
    elif distance <= np.absolute(radius1 - radius2):
        if full_area1 > full_area2:
            return (full_area1, 0)
        if full_area1 < full_area2:
            return (0, full_area2)
        else:
            return (0.5 * full_area1, 0.5 * full_area2)
    else:
        if radius1 > 0:
            cos_theta1 = (rad1sq + distsq - rad2sq) / (2 * radius1 * distance)
            cap1_area = 2 * np.pi * radius1 * radius1 * (1 - cos_theta1)
        else:
            cap1_area = 0
        if radius2 > 0:
            cos_theta2 = (rad2sq + distsq - rad1sq) / (2 * radius2 * distance)
            cap2_area = 2 * np.pi * radius2 * radius2 * (1 - cos_theta2)
        else:
            cap2_area = 0
        return (full_area1 - cap1_area, full_area2 - cap2_area)


@pytest.mark.parametrize("radius", [0.0, 1.1, 2.2, 4.4, 6.6, 8.8])
def test_two_sphere_sasa(radius, tmp_path):
    """Test solvent accessible surface areas for two spheres."""
    atom_tolerance = 0.02
    total_tolerance = 0.02
    probe_radius = 0.0
    big_atom = Atom()
    big_atom.radius = radius
    big_atom.position = np.array([0, 0, 0])
    little_atom = Atom()
    little_atom.radius = 1.0
    test_atom_areas = []
    test_total_areas = []
    ref_atom_areas = []
    ref_total_areas = []
    distances = np.linspace(0, (big_atom.radius + little_atom.radius), num=20)
    for distance in distances:
        _LOGGER.debug("Distance = %g", distance)
        little_atom.position = np.array(3 * [distance / np.sqrt(3)])
        xyz_path = Path(tmp_path) / f"spheres-{distance}.xyz"
        sas = SolventAccessibleSurface(
            [big_atom, little_atom], probe_radius, 1000, xyz_path=xyz_path
        )
        test = np.array([sas.atom_surface_area(0), sas.atom_surface_area(1)])
        test_total_areas.append(test.sum())
        test_atom_areas.append(test)
        ref = np.array(
            two_sphere_area(big_atom.radius, little_atom.radius, distance)
        )
        ref_total_areas.append(ref.sum())
        ref_atom_areas.append(ref)
    test_atom_areas = np.array(test_atom_areas)
    test_total_areas = np.array(test_total_areas)
    ref_atom_areas = np.array(ref_atom_areas)
    ref_total_areas = np.array(ref_total_areas)
    rel_difference = np.absolute(
        np.divide(test_atom_areas - ref_atom_areas, np.sum(ref_atom_areas))
    )
    errors = []
    if np.any(rel_difference > atom_tolerance):
        ref_series = pd.Series(index=distances, data=ref_total_areas)
        ref_series.index.name = "Dist"
        ref_df = pd.DataFrame(index=distances, data=ref_atom_areas)
        ref_df.index.name = "Dist"
        test_df = pd.DataFrame(index=distances, data=test_atom_areas)
        test_df.index.name = "Dist"
        abs_diff_df = pd.DataFrame(
            index=distances,
            data=np.absolute(ref_atom_areas - test_atom_areas),
        )
        abs_diff_df.index.name = "Dist"
        rel_diff_df = pd.DataFrame(index=distances, data=rel_difference)
        rel_diff_df.index.name = "Dist"
        _LOGGER.info(
            f"\nTolerance exceeded: {atom_tolerance}\n"
            f"\nRadii: {little_atom.radius}, {big_atom.radius}\n"
            f"\nReference total areas:\n{ref_series}\n"
            f"\nReference atom areas:\n{ref_df}\n"
            f"\nTest atom areas:\n{test_df}\n"
            f"\nAbsolute differences:\n{abs_diff_df}\n"
            f"\nRelative differences:\n{rel_diff_df}\n"
        )
        errors.append(
            "Atomic surface area relative differences exceed tolerance (%g)"
            % atom_tolerance
        )
    rel_differences = np.absolute(
        np.divide(test_total_areas - ref_total_areas, ref_total_areas)
    )
    if np.any(rel_differences > total_tolerance):
        ref_series = pd.Series(index=distances, data=ref_total_areas)
        ref_series.index.name = "Dist"
        test_series = pd.Series(index=distances, data=test_total_areas)
        test_series.index.name = "Dist"
        abs_series = pd.Series(
            index=distances,
            data=np.absolute(ref_total_areas - test_total_areas),
        )
        abs_series.index.name = "Dist"
        rel_series = pd.Series(index=distances, data=rel_differences)
        rel_series.index.name = "Dist"
        _LOGGER.error(
            f"\nTolerance exceeded:  {total_tolerance}\n"
            f"\nRadii: {little_atom.radius}, {big_atom.radius}\n"
            f"\nReference areas:\n{ref_series}\n"
            f"\nTest areas:\n{test_series}\n"
            f"\nAbsolute differences:\n{abs_series}\n"
            f"\nRelative differences:\n{rel_series}"
        )
        errors.append(
            "Total surface area relative differences exceed tolerance (%g)"
            % total_tolerance
        )
    if errors:
        raise AssertionError(";".join(errors))


@pytest.mark.parametrize("molecule", list(ATOM_AREAS.keys()))
def test_atom_sasa(molecule, tmp_path):
    """Test per-atom solvent-accessible surface areas for small molecules."""
    _LOGGER.info("Temp path:  %s", tmp_path)
    atom_tolerance = 0.03
    total_tolerance = 0.03
    pqr_path = "%s.pdb" % molecule
    pqr_path = Path("tests/data/alkanes") / pqr_path
    with open(pqr_path, "rt") as pqr_file:
        atoms = parse_pqr_file(pqr_file)
    xyz_path = Path(tmp_path) / f"{molecule}.xyz"
    sas = SolventAccessibleSurface(atoms, 0.65, 4574, xyz_path=xyz_path)
    test_areas = np.array(
        [sas.atom_surface_area(iatom) for iatom in range(len(atoms))]
    )
    test_areas = test_areas
    ref_areas = np.array(ATOM_AREAS[molecule])
    ref_areas = ref_areas
    abs_diff = np.absolute(test_areas - ref_areas)
    rel_diff = abs_diff / np.sum(ref_areas)
    keys = []
    for atom in atoms:
        keys.append(f"{atom.pqr_atom_num} {atom.atom_name}")
    df = pd.DataFrame(
        index=keys,
        data={
            "Ref values": ref_areas,
            "Test values": test_areas,
            "Abs diff": abs_diff,
            "Rel diff": rel_diff,
        },
    )
    _LOGGER.info("\n%s", df.sort_values("Abs diff", ascending=False))
    errors = []
    if np.any(rel_diff > atom_tolerance):
        _LOGGER.error("\n%s", df.sort_values("Rel diff", ascending=False))
        errors.append(
            "Per-atom relative differences exceed tolerance (%g)"
            % atom_tolerance
        )
    abs_error = np.absolute(np.sum(test_areas) - np.sum(ref_areas))
    rel_error = abs_error / np.sum(ref_areas)
    if rel_error > total_tolerance:
        _LOGGER.warning(
            f"Tolerance exceeded: {total_tolerance}\n"
            f"Reference area: {np.sum(ref_areas)}\n"
            f"Test area: {np.sum(test_areas)}\n"
            f"Absolute difference: {abs_error}\n"
            f"Relative difference: {rel_error}\n"
        )
        errors.append(
            "Total relative difference exceeds tolerance (%g)"
            % total_tolerance
        )
    if errors:
        raise AssertionError(";".join(errors))


@pytest.mark.parametrize(
    "pqr_path,ref_json", [("NikR_chains.pqr", "NikR.json")]
)
def test_protein_details(pqr_path, ref_json, tmp_path):
    """Test SASA performance for proteins."""
    _LOGGER.info("Temp path:  %s", tmp_path)
    abs_cutoff = 2.5
    num_abs_cutoff = 26
    rel_cutoff = 0.1
    num_rel_cutoff = 33
    tot_abs_cutoff = 24.0
    tot_rel_cutoff = 0.002
    pqr_path = PROTEIN_PATH / pqr_path
    ref_json = PROTEIN_PATH / ref_json
    with open(pqr_path, "rt") as pqr_file:
        atoms = parse_pqr_file(pqr_file)
    xyz_path = Path(tmp_path) / "surface.xyz"
    sas = SolventAccessibleSurface(
        atoms, probe_radius=1.4, num_points=5026, xyz_path=xyz_path
    )
    test_areas = [sas.atom_surface_area(iatom) for iatom in range(len(atoms))]
    df = aggregate(
        atoms,
        test_areas,
        chain_id=True,
        res_name=False,
        res_num=True,
        sidechain=False,
    )
    with open(ref_json, "rt") as ref_file:
        ref_dict = json.load(ref_file)
    keys = []
    ref_values = []
    test_values = []
    for chain in ref_dict.keys():
        chain_df = df[df["chain_id"] == chain]
        test_values = test_values + list(chain_df["value"].values)
        keys = keys + [f"{chain} {key}" for key in ref_dict[chain].keys()]
        ref_values = ref_values + list(ref_dict[chain].values())
    ref_values = np.array(ref_values)
    test_values = np.array(test_values)
    abs_diff = np.absolute(ref_values - test_values)
    rel_diff = np.divide(abs_diff, ref_values)
    df = pd.DataFrame(
        index=keys,
        data={
            "Ref values": ref_values,
            "Test values": test_values,
            "Abs diff": abs_diff,
            "Rel diff": rel_diff,
        },
    )
    _LOGGER.info(
        "Largest absolute differences:\n%s", df.nlargest(20, "Abs diff")
    )
    df = df.dropna()
    _LOGGER.info(
        "Largest relative differences:\n%s", df.nlargest(20, "Rel diff")
    )
    errors = []
    num_abs = len(abs_diff[abs_diff > abs_cutoff])
    if num_abs > num_abs_cutoff:
        err = (
            f"Number of absolute differences ({num_abs}) above {abs_cutoff} "
            f"exceeds limit ({num_abs_cutoff})"
        )
        errors.append(err)
    num_rel = len(rel_diff[rel_diff > rel_cutoff])
    if num_rel > num_rel_cutoff:
        err = (
            f"Number of relative differences ({num_rel}) above {rel_cutoff}) "
            f"exceeds limit ({num_rel_cutoff})"
        )
        errors.append(err)
    ref_total = np.sum(ref_values)
    test_total = np.sum(test_values)
    abs_total = np.absolute(ref_total - test_total)
    if abs_total > tot_abs_cutoff:
        err = (
            f"Absolute difference in total area ({abs_total}) exceeds "
            f"limit ({tot_abs_cutoff})"
        )
        errors.append(err)
    rel_total = abs_total / ref_total
    if rel_total > tot_rel_cutoff:
        err = (
            f"Relative difference in total area ({rel_total}) exceeds "
            "limit ({tot_rel_cutoff})"
        )
        errors.append(err)
    if errors:
        raise AssertionError(";".join(errors))


@pytest.mark.parametrize("protein", ["1A6F", "1STN", "2BU4"])
def test_auton_sasa(protein, tmp_path):
    """Test aggregate SASAs per residue type as reported in Auton and Bolen
    (doi:10.1073/pnas.0507053102, Supporting Table 2)."""
    _LOGGER.info("Temp path:  %s", tmp_path)
    pqr_path = PROTEIN_PATH / f"{protein}.pqr"
    with open(pqr_path, "rt") as pqr_file:
        atoms = parse_pqr_file(pqr_file)
    xyz_path = Path(tmp_path) / "surface.xyz"
    sas = SolventAccessibleSurface(
        atoms, probe_radius=1.4, num_points=5000, xyz_path=xyz_path
    )
    test_areas = [sas.atom_surface_area(iatom) for iatom in range(len(atoms))]
    test_areas = aggregate(
        atoms,
        test_areas,
        chain_id=False,
        res_name=True,
        res_num=False,
        sidechain=True,
    )
    test_areas = test_areas.set_index("res_name").sort_index()
    test_counts = np.array(count_residues(atoms))
    ref_path = PROTEIN_PATH / "Auton-Bolen.yaml"
    with open(ref_path, "rt") as ref_file:
        ref_dict = yaml.load(ref_file, Loader=yaml.FullLoader)
    ref_dict = ref_dict[protein]
    ref_df = pd.DataFrame(ref_dict).T
    ref_df = ref_df[ref_df["number"] > 0]
    ref_counts = np.array(ref_df["number"].sort_index())
    np.testing.assert_equal(test_counts, ref_counts)
    for group in ["backbone", "sidechain"]:
        ref = np.array(ref_df[group])
        test = np.array(test_areas[test_areas["sidechain"] == group]["value"])
        # Test correlation since we're not 100% sure what parameters were used
        # in PNAS paper
        results = linregress(ref, test)
        info = (
            f"{group} areas are correlated with results from paper:  "
            f"{results}"
        )
        _LOGGER.info(info)
        if not np.isclose(results.pvalue, 0):
            err = f"Poor fit to {group} areas:  {results}"
            raise AssertionError(err)


@pytest.mark.parametrize(
    "amino_acid",
    [
        "ALA",
        "GLY",
        "ILE",
        "LEU",
        "PRO",
        "VAL",
        "PHE",
        "TRP",
        "TYR",
        "ASP",
        "GLU",
        "ARG",
        "HIS",
        "LYS",
        "SER",
        "THR",
        "CYS",
        "MET",
        "ASN",
        "GLN",
    ],
)
def test_unfolded_sasa(amino_acid):
    """Test unfolded SASA models by comparing results of Auton and Bolen
    (doi:10.1073/pnas.0507053102) with Creamer et al
    (doi:10.1021/bi962819o)."""
    err_tol = 0.101
    ref_models = ReferenceModels()
    val1 = pd.Series(ref_models.residue_area(amino_acid, model="auton"))
    val2 = pd.Series(ref_models.residue_area(amino_acid, model="creamer"))
    print(val1, val2)
    abs_err = np.absolute(val1 - val2)
    rel_err = 2 * abs_err / (val1 + val2)
    diag_str = (
        f"{amino_acid}:\nAuton:\n{val1}\nCreamer:\n{val2}\n"
        f"Abs error:\n{abs_err}\nRel error:\n{rel_err})\n"
    )
    _LOGGER.info(diag_str)
    # 0.10 accounts for rounding error between the two papers
    if np.any(abs_err > err_tol):
        if amino_acid != "PHE":
            err = f"Error tolerance {err_tol} exceeded for {diag_str}"
            raise AssertionError(err)
        _LOGGER.error(f"{diag_str}\nError in Auton value?")


tests/test_energy.py
"""Test m-value prediction model."""
import logging
from pathlib import Path
import pandas as pd
import numpy as np
from scipy.stats import linregress
from osmolytes.main import main


_LOGGER = logging.getLogger(__name__)
PROTEIN_PATH = Path("tests/data/proteins")
# Results (kcal/mol) were extracted visually from the Figure 1 in the paper
# doi: 10.1073/pnas.0507053102
AUTON_RESULTS = {
    "2BU4 urea": -2.40,
    "2BU4 betaine": -0.25,
    "2BU4 proline": -0.05,
    "2BU4 sorbitol": 0.80,
    "2BU4 sucrose": 1.00,
    "2BU4 sarcosine": 1.5,
    "2BU4 TMAO": 1.80,
    "1STN TMAO": 2.85,
    "1A6F TMAO": 3.00,
}


def test_energy(tmp_path):
    """Test the energy model as reported in Auton and Bolen
    (doi:10.1073/pnas.0507053102, Figure 1)"""
    test_results = {}
    for protein in ["1A6F", "1STN", "2BU4"]:
        pqr_path = PROTEIN_PATH / f"{protein}.pqr"
        xyz_path = Path(tmp_path) / f"{protein}.xyz"
        results = main(
            [
                "--solvent-radius",
                "1.4",
                "--surface-points",
                "2000",
                "--surface-output",
                str(xyz_path),
                str(pqr_path),
            ]
        )
        energy_df = results["energy_df"]
        _LOGGER.info(f"{protein} detailed energies:\n{energy_df.to_string()}")
        energies = results["energies"]
        _LOGGER.info(f"{protein} m-values\n{energies}")
        for osmolyte, value in energies.iteritems():
            key = f"{protein} {osmolyte}"
            if key in AUTON_RESULTS:
                test_results[key] = value
    ref_results = pd.Series(AUTON_RESULTS).sort_index()
    test_results = pd.Series(test_results).sort_index()
    abs_error = np.absolute(ref_results - test_results)
    rel_error = np.divide(abs_error, np.absolute(ref_results))
    err_df = pd.DataFrame(
        {
            "Ref results": ref_results,
            "Test results": test_results,
            "Abs error": abs_error,
            "Rel error": rel_error,
        }
    )
    _LOGGER.info(f"Test results:\n{err_df}")
    results = linregress(ref_results, test_results)
    info = f"Correlation of test with results from paper:  {results}"
    _LOGGER.info(info)
    if not np.isclose(results.pvalue, 0):
        err = f"Poor fit to predicted energies:  {results}"
        raise AssertionError(err)


osmolytes/main.py
"""Main driver for module."""
import argparse
import logging
from pathlib import Path
from sys import version_info
import osmolytes
from osmolytes.pqr import parse_pqr_file, count_residues
from osmolytes.sasa import SolventAccessibleSurface
from osmolytes.energy import transfer_energy


assert version_info > (3, 5)
_LOGGER = logging.getLogger("main")


def build_parser():
    """Build argument parser.

    :returns:  argument parser
    :rtype:  argparse.ArgumentParser
    """
    parser = argparse.ArgumentParser(
        description=(
            f"Osmolytes v. {osmolytes.__version__}. "
            f"This code uses various models to attempt to predict the "
            f"influence of osmolytes on protein stability"
        ),
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument(
        "--log-level",
        help="set verbosity of code",
        nargs=1,
        default="INFO",
        choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"],
    )
    parser.add_argument(
        "--model",
        help=(
            "model to use for stability prediction:\n"
            "  'auton' -- Auton M, Bolen DW. Predicting the energetics of "
            "osmolyte-induced protein folding/unfolding. Proc Natl Acad Sci "
            "102:15065 (2005) https://doi.org/10.1073/pnas.0507053102"
        ),
        nargs=1,
        default="auton",
        choices=["auton"],
    )
    parser.add_argument(
        "--surface-output",
        default=None,
        help="path for output of surface in XYZ format",
    )
    parser.add_argument(
        "--solvent-radius",
        default=1.4,
        type=float,
        help="radius of solvent molecule (in Å)",
    )
    parser.add_argument(
        "--surface-points",
        default=2000,
        type=int,
        help="number of points per atom for constructing surface",
    )
    parser.add_argument(
        "--output",
        choices=["xlsx", "csv"],
        help="output m-value results in Excel (xlsx) or CSV (csv) format",
    )
    parser.add_argument(
        "--output-dir", default=".", help="directory for m-value output"
    )
    parser.add_argument(
        "pqr_path",
        help=(
            "input PQR file for protein, as generated by PDB2PQR or "
            "similar program (see documentation) "
        ),
    )
    return parser


def main(args=None):
    """Main entry point.

    :param list(str) args:  list of strings to parse
    :returns:  dictionary of results
    :rtype:  dict
    """
    parser = build_parser()
    args = parser.parse_args(args)
    logging.basicConfig(level=getattr(logging, args.log_level))
    _LOGGER.info(f"Osmolytes version {osmolytes.__version__}")
    _LOGGER.debug(f"Got command-line arguments: {args}")
    _LOGGER.info(f"Reading PQR input from {args.pqr_path}...")
    with open(args.pqr_path, "rt") as pqr_file:
        atoms = parse_pqr_file(pqr_file)
    count_df = count_residues(atoms)
    _LOGGER.info(f"Protein composition:\n{count_df}")
    _LOGGER.info(
        f"Constructing protein solvent-accessible surface with "
        f"solvent radius {args.solvent_radius} and {args.surface_points} "
        f"points per atom."
    )
    sas = SolventAccessibleSurface(
        atoms,
        probe_radius=args.solvent_radius,
        num_points=args.surface_points,
        xyz_path=args.surface_output,
    )
    _LOGGER.info("Calculating transfer energies.")
    energy_df = transfer_energy(atoms, sas)
    _LOGGER.info(f"Detailed energies (kcal/mol/M):\n{energy_df.to_string()}")
    energies = energy_df.sum(axis=0).sort_values()
    _LOGGER.info(f"Summary energies (kcal/mol/M):\n{energies}")
    output_dir = Path(args.output_dir)
    if args.output == "xlsx":
        output_path = output_dir / f"{Path(args.pqr_path).stem}-mvalues.xlsx"
        _LOGGER.info(f"Writing energies (kcal/mol/M) to {output_path}")
        energy_df.to_excel(output_path)
    elif args.output == "csv":
        output_path = output_dir / f"{Path(args.pqr_path).stem}-mvalues.csv"
        _LOGGER.info(f"Writing energies (kcal/mol/M) to {output_path}")
        energy_df.to_csv(output_path)
    return {
        "args": args,
        "atoms": atoms,
        "sas": sas,
        "energy_df": energy_df,
        "energies": energies,
    }


def console():
    """Run code from the console."""
    main()
